Решение финального задания 15 спринта (add-bookings)

RecoilSPB opened the pull request.

Recoilvrn committed:
Решение финального задания 15 спринта (add-bookings)

z8q commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -0,0 +1,36 @@
+import java.time.LocalDateTime;
+
+
+@Data
```
На дипломе будут рекомендовать не использовать Data с сущностями,
статьи про это:
https://thorben-janssen.com/lombok-hibernate-how-to-avoid-common-pitfalls/
https://habr.com/ru/companies/haulmont/articles/564682/
https://jpa-buddy.com/blog/hopefully-the-final-article-about-equals-and-hashcode-for-jpa-entities-with-db-generated-ids/

Один из примеров возможных проблем
Data содержит в себе toString

У нас например есть сущность с каким-нибудь полем `OneToMany Set<Item> items`,
У OneToMany по умолчанию параметр fetch == FetchType.LAZY
Это значит, что, если мы запросим из базы саму сущность, items не будут подгружены,
но у нас есть созданный ломбоком toString(), и если мы где-то его вызовем, то могут возникать дополнительные обращения к базе
Если такие обращения будут при уже закрытой сессии хибернейта, то можем получить LazyInitializationException

z8q commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceDb.java:
```diff
@@ -0,0 +1,182 @@
+        this.bookingRepository = bookingRepository;
+        this.userRepository = userRepository;
+        this.itemRepository = itemRepository;
+    }
```
В строках 29-31 все поля final и non-initialized, можно вместо такого конструктора с Autowired поставить аннотацию RequiredArgsConstructor над классом

z8q commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceDb.java:
```diff
@@ -0,0 +1,182 @@
+                .map(BookingMapper::toBookingDto)
+                .sorted(new BookingComparator().reversed())
+                .collect(Collectors.toList());
+    }
```
По логике, метод мапит лист с объектами одного типа в лист с объектами другого типа
Можно перенести в маппер

z8q commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceDb.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    @Override
+    public BookingOutputDto updateApprove(Long bookingId, Boolean approved, Long userId) {
```
Методы, которые сохраняют записи в бд можно обернуть в транзакции, поставить над методом Transactional
В методах, где происходит только чтение, можно еще указать параметр readOnly = true

Еще есть момент, что транзакции будут создаваться, только если вызовы метода происходят вне класса.
Если один метод внутри класса будет вызывать другой, помеченный Transactional, то в этом случае транзакция не будет создаваться. Можно про это почитать, либо можешь написать мне)

z8q commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceDb.java:
```diff
@@ -0,0 +1,182 @@
+        List<Item> items = itemRepository.findByOwner(owner);
+        log.info("Получен запрос на получение списка бронирований по владельцу");
+        return items.stream()
+                .flatMap((item) -> getAllBookingsForItem(item, state).stream())
```
тут скорее всего будет сложно избежать обращения к базе в цикле,
но на дипломе это будут смотреть,
нужно будет стараться избегать обращений в циклах и стримах, передавать сразу список id например и получать результат тоже списком. Сохранять списками через saveAll и т.д.

При уменьшении количества обращений к базе, скорость работы программы будет выше

Recoilvrn committed:
Перенес метод convertBookings в класс BookingMapper
Заменил конструктор на анотацию @RequiredArgsConstructor

Recoilvrn committed:
Избавился от циклического обращения к БД

Recoilvrn committed:
Добавил анотацию Transactional для методов

Recoilvrn committed:
Заменил анотацию @Data на @Getter и @Setter
Добавил анотацию @EqualsAndHashCode(onlyExplicitlyIncluded = true) - с указанием только значимых полей, чтобы избежать циклических ссылок и ненужных запросов к базе, @ToString — с исключением ленивых коллекций через параметр exclude

RecoilSPB committed:
fix Checkstyle
